packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param symmetricKey
+ */
+async function fetchOrGenerateKey(
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  symmetricKey?: Buffer
+) {
+  if (symmetricKey) {
+    return Ok(symmetricKey)
+  }
+
+  const existingKey = await readSymmetricKey(wrapper, dataPath, wrapper.self)
+  if (existingKey.ok) {
+    return Ok(existingKey.result)
+  }
+
+  if (
+    existingKey.error.errorType === SchemaErrorTypes.OffchainError &&
+    existingKey.error.error.errorType === OffchainErrorTypes.NoStorageRootProvidedData
+  ) {
+    return Ok(randomBytes(16))
+  }
+
+  return Err(existingKey.error)
+}
+
+/**
+ * Handles encrypting the data with a symmetric key, then distributing said key to each address
+ * in the `toAddresses` array.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param data the data to encrypt
+ * @param toAddresses the addresses to distribute the symmetric key to
+ * @param symmetricKey the symmetric key to use to encrypt the data. One will be found or generated if not provided
+ */
+export const writeEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  data: Buffer,
+  toAddresses: string[],
+  symmetricKey?: Buffer
+) => {
+  const fetchKey = await fetchOrGenerateKey(wrapper, dataPath, symmetricKey)
+  if (!fetchKey.ok) {
+    return fetchKey.error
+  }
+
+  //  file_ciphertext = IV || E(message key, IV, data)
+  const iv = randomBytes(16)
+  const cipher = createCipheriv('aes-128-ctr', fetchKey.result, iv)
+  const payload = Buffer.concat([iv, cipher.update(data), cipher.final()])
+
+  const signature = await signBuffer(wrapper, `${dataPath}.enc`, payload)
+  await wrapper.writeDataTo(payload, signature, `${dataPath}.enc`)
+
+  await Promise.all(
+    // here we encrypt the key to ourselves so we can retrieve it later
+    [wrapper.self, ...toAddresses].map(async (toAddress) =>
+      distributeSymmetricKey(wrapper, dataPath, fetchKey.result, toAddress)
+    )
+  )
+}
+
+/**
+ * Reads and decrypts a symmetric key that has been encrypted to your
+ * data encryption key.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+const readSymmetricKey = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const accounts = await wrapper.kit.contracts.getAccounts()
+  const wallet = wrapper.kit.getWallet()
+  const [readerPubKey, senderPubKey] = await Promise.all([
+    accounts.getDataEncryptionKey(wrapper.self),
+    accounts.getDataEncryptionKey(senderAddress),
+  ])
+
+  if (readerPubKey === null) {
```

Comment:
Should this be `senderPubKey`? If so, the a check may need to be added for `readerPubKey` before sending it to `publicKeyToAddress`